code/src/main/kotlin/com/expediagroup/sdk/core/http/Url.kt

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (C) 2024 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.sdk.core.http
+
+import com.expediagroup.sdk.core.http.Url.Companion.parse
+import java.net.IDN
+import java.net.InetAddress
+import java.net.MalformedURLException
+import java.net.URI
+import java.net.URISyntaxException
+import java.net.URL
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+import java.util.Locale
+
+/**
+ * Represents an immutable URL with proper encoding and parsing.
+ *
+ * The [Url] class provides functionality for creating, validating, and manipulating URLs. It ensures that
+ * URLs are correctly encoded, parsed, and represented as standardized strings or objects like [URI] and [URL].
+ *
+ * Instances of [Url] are immutable and must be created using the [Builder] or the [parse] method.
+ */
+class Url private constructor(
```

Comment:
Why don't we use URL/URI class from Java?

Reply:
This class adds a builder methods and handles more encoding, Ipv6, validations concerns on the URLs. I'll revisit it and see if it worths keeping it or just replace with the native classes